src/plugins/visualizations/public/embeddable/visualize_embeddable.tsx

```diff
@@ -233,13 +266,22 @@ export class VisualizeEmbeddable extends Embeddable<VisualizeInput, VisualizeOut
 
     this.transferCustomizationsToUiState();
 
+    this.labelNode = document.createElement('div');
+    this.labelNode.className = `panel-label`;
+    domNode.appendChild(this.labelNode);
+
     const div = document.createElement('div');
     div.className = `visualize panel-content panel-content--fullWidth`;
     domNode.appendChild(div);
+
     this.domNode = div;
 
     const expressions = getExpressions();
-    this.handler = new expressions.ExpressionLoader(this.domNode);
+    this.handler = new expressions.ExpressionLoader(this.domNode, '', {
+      onRenderError: (element: HTMLElement, error: RenderError) => {
+        this.onContainerError(error);
```

Comment:
I haven't test yet, but I wonder:

considering very recent fixes around race conditions in this file, could this callback become another source for possible race condition? 

Like for example: there is already a new request finished and new vis rendered, but somehow error from previous request popped up and replaced actual visualization with older error? 

There are some checks happening in `updateHander`, but it relies on reference to older abort controller.

Reply:
This callback is called exactly the same way it was previously - just with different UI logic.

See `src/plugins/expressions/public/render.ts` line 68.

We used to use `defaultRenderErrorHandler` before, and it would output the errors as toasts. Since we now pass a custom error handler - it gets called instead when an error occurs.

There are no other changes in flow. I just extracted the state handling into meaningful callbacks to make it more readable.

Reply:
I think there is a slight difference: 

1. Previously we just showed a popup on the right and error didn't affect the visualisation embeddable.
2. Now if older error pops up we will show "error" label near the visualisation in actual state.

I still haven't reproduced what I am worrying about there :( 
Just being extra cautions because of recent experience with cancelations :(
Don't have a particular action item here now